### PR TITLE
#83 Add Conformance Rule dialog validation

### PR DESCRIPTION
**Pull request by GeorgiChochov**
Validates:
* all required fields in the conformance rule forms
* minimum of 2 columns for concatenation
* concatenation column dialog
* minimum of 1 join condition for mapping
* join condition dialog
* output column that might override (of if inserting a rule in the middle of the list, be overridden by another rule later on)
* drop only columns which do not break the transitive schema (if adding a drop rule in the middle of the list)

### CODE REVIEW

**Comment by Zejnilovic** [comment]
## 1 Unable to create nested structures
I am unable to create nested structures. For example, going to concatenation rule and in `OutputColumn` putting `alfa.beta`

image

File: menas/ui/components/dataset/datasetDetail.controller.js [hunk 1]
```diff
@@ -325,16 +325,21 @@ sap.ui.define([
...
-
-      	const auditTable = this.byId("auditTrailTable");
-      	this._datasetService.getAuditTrail(currentDataset.name, auditTable);
+        this._transitiveSchemas = [];
+        const transitiveSchemas = this._transitiveSchemas;
+        new SchemaRestDAO().getByNameAndVersionSync(currentDataset.schemaName, currentDataset.schemaVersion).then((schema) => {
+          this._model.setProperty("/currentDataset/schema", schema)
+          this._schemaTable.model = schema;
+          transitiveSchemas.push(schema);
+          // var millisecondsToWait = 5000;
```
Comment (lokm01) [inline menas/ui/components/dataset/datasetDetail.controller.js]: Maybe clean these up?

**Comment by Zejnilovic** [comment]
## 2 Input column alias in SingleColmn Rule allows `.`
Input column alias in SingleColmn Rule allows `.` which then does not translate to the embedded structure but as a name itself which can be misleading and lead to errors in the end. Haven't yet run 2e2 using this rule in this setting.

**Conformance rule view**
image

**Schema picker**
image

**Conformance rule edit**
image

File: menas/ui/components/dataset/conformanceRule/ConformanceRule.js [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+}
+
+class ConcatenationConformanceRule extends ConformanceRule {
+
+  constructor(rule) {
+    super(rule);
+  }
+
+  apply(fields) {
+    const isNullable = this.getInputCols(fields).some(field => field.nullable);
```
[the remark this answers is not in the document]
Comment (GeorgiChochov) [reply menas/ui/components/dataset/conformanceRule/ConformanceRule.js]: Nope, if one is nullable, the entire thing becomes nullable

File: menas/ui/components/dataset/conformanceRule/ConformanceRuleForm.js [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+}
+
+class DropConformanceRuleForm extends ConformanceRuleForm {
+
+  constructor() {
+    super("DropConformanceRule", true)
+  }
+
+  get outputColumnControl() {
+    return sap.ui.getCore().byId("DropConformanceRule--schemaFieldSelector");
```
[the remark this answers is not in the document]
Comment (GeorgiChochov) [reply menas/ui/components/dataset/conformanceRule/ConformanceRuleForm.js]: Left over from when it had a reason.

File: menas/ui/components/dataset/conformanceRule/ConformanceRuleForm.js [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+    return this.hasValidSchemaSelection(fieldValue, this.outputColumnControl);
+  }
+
+  reset() {
+    this.outputColumnControl
+      .getItems()
+      .forEach(item => item.setHighlight(sap.ui.core.ValueState.None));
+  }
+
+  hasValidTransitiveSchema(rule, schemas, rules) {
```
[the remark this answers is not in the document]
Comment (GeorgiChochov) [reply menas/ui/components/dataset/conformanceRule/ConformanceRuleForm.js]: I think we don't. Unless something uses `level1.level2` there's no reason to worry about it.

File: menas/ui/components/dataset/conformanceRule/ConformanceRuleForm.js [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+  }
+
+  isValid(rule, schemas, rules) {
+    const hasValidOutputColumn = this.hasValidOutputColumn(rule.outputColumn)
+      && this.hasValidTransitiveSchema(rule, schemas, rules);
+    return hasValidOutputColumn & this.isCorrectlyConfigured(rule);
+  }
+
+  isCorrectlyConfigured(rule) {
+    return true;
```
[the remark this answers is not in the document]
Comment (GeorgiChochov) [reply menas/ui/components/dataset/conformanceRule/ConformanceRuleForm.js]: Good spot, this is left over from initial setup.
As for the abstract method, I already did that for `ConformanceRule::apply` a while ago, and will do it here as well.

File: menas/ui/service/GenericService.js [hunk 6]
```diff
@@ -66,10 +66,18 @@ var GenericService = new function () {
     })
   };
 
+  this.isEmpty = function (str) {
+    return !str || str === "";
```
Comment (Zejnilovic) [inline menas/ui/service/GenericService.js]: The second part is unneeded I think

image

**Comment by Zejnilovic** [comment]
## 3 Uncaught UseCase
Steps:
1. Create a rule with output column `Alfa`
2. Create a drop rule for `Alfa`
3. Create another rule with output column `Alfa`
4. Move rule from step 3. above the drop rule.

This is something unexpected and should not happen as now I have two rules that create output column `Alfa`.

**Comment by Zejnilovic** [comment]
## 1 Unable to create nested structures

Currently, I get the error below.
image

File: menas/ui/service/GenericService.js (hunk 6 above)
Comment (GeorgiChochov) [reply menas/ui/service/GenericService.js]: Holy crap, JS takes a lot of liberties in interpreting things as a boolean lol

**Comment by GeorgiChochov** [comment]
> ## 1 Unable to create nested structures
> Currently, I get the error below.
> image

Pushed the fixed for that a minute ago, when I noticed it as well.

**Comment by GeorgiChochov** [comment]
> ## 3 Uncaught UseCase
> Steps:
> 
> 1. Create a rule with output column `Alfa`
> 2. Create a drop rule for `Alfa`
> 3. Create another rule with output column `Alfa`
> 4. Move rule from step 3. above the drop rule.
> 
> This is something unexpected and should not happen as now I have two rules that create output column `Alfa`.

I remember specifically fixing this... Had it in mind since the rule reordering implementation.

File: menas/ui/components/dataset/conformanceRule/ConformanceRuleForm.js [hunk 7]
```diff
@@ -0,0 +1,423 @@
...
+  }
+
+  hasValidInputColumn(fieldValue) {
+    return this.hasValidSchemaSelection(fieldValue, this.inputColumnControl);
+  }
+
+  hasValidSchemaSelection(fieldValue, inputControl) {
+    let isValid = true;
+
+    if (!fieldValue || fieldValue === "" || !GenericService.isValidColumnName(fieldValue)) {
```
Comment (Zejnilovic) [inline menas/ui/components/dataset/conformanceRule/ConformanceRuleForm.js]: Here is the `!fieldValue` enough as well.

File: menas/ui/components/dataset/conformanceRule/ConformanceRuleForm.js [hunk 8]
```diff
@@ -0,0 +1,423 @@
...
+  isCorrectlyConfigured(rule) {
+    return this.hasValidInputColumn(rule.inputColumn)
+      & this.hasValidOutputDataType(rule.outputDataType);
+  }
+
+  hasValidOutputDataType(fieldValue) {
+    const inputControl = this.outputDataTypeControl;
+    let isValid = true;
+
+    if (!fieldValue || fieldValue === "") {
```
Comment (Zejnilovic) [inline menas/ui/components/dataset/conformanceRule/ConformanceRuleForm.js]: Here is the `!fieldValue` enough as well.

**Comment by Zejnilovic** [comment]
## 1 Unable to create nested structures
**Fixed**

## 2 Input column alias in SingleColmn Rule allows `.`
**Fixed**

**Comment by GeorgiChochov** [comment]
> ## 3 Uncaught UseCase
> Steps:
> 
> 1. Create a rule with output column `Alfa`
> 2. Create a drop rule for `Alfa`
> 3. Create another rule with output column `Alfa`
> 4. Move rule from step 3. above the drop rule.
> 
> This is something unexpected and should not happen as now I have two rules that create output column `Alfa`.

This PR is too big as it is, let's build on in separate tasks. Created #532 for this.

**Pull request merged**